### ISSUE TITLE
Allow Optional Inputs

### ISSUE THREAD

**Issue by elijahbenizzy** [DAGWorks-Inc/burr#176]
**Is your feature request related to a problem? Please describe.**
If you want multiple input types, this is a clean way to do it. This currently doesn't work:

```python
@action(...)
def process_input(state: State, input_text: str, input_doc=None: Document=None) -> ...:
    ...
```

This is something like what chatGPT does.

**Describe the solution you'd like**
This should work. It should make sure not to break if its not provided. Currently this will:

1. Not parse it as an input: see https://github.com/DAGWorks-Inc/burr/blob/26cfcd272a26c58e88c6ec84b18682b8ad87a9bd/burr/core/action.py#L370
2. Then not process it -- meaning it will break silently...

Instead, we should:
1. Enable a list of optional inputs
2. Break only if all required inputs are present

Should be as easy as adding another + some tests.

**Describe alternatives you've considered**

This is generally solvable by grouping into an input type that encodes the default relationships:

```python
class UserInput(pydantic.BaseModel):
    prompt: Optional[str]
    document: Optional[Document]

@action(reads=..., writes=["document", "prompt"])
def prompt(state: State, user_input: UserInput):
    ...
```

This has the advantage of being easier to manage/inspect from a single object in the UI. Furthermore, any more complex
relationships can be easily encoded here. That said, it's less ergonomic, and requires some sort of object/dataclass setup.

**Additional context**
Inspired by user Milan on discord!

**Comment by AshisGhosh** [comment]
Adding to this - 

I have a use case where I may which to pass optional parameters to a prompt generation node. The prompt generated would depend on which parameters are available. The workaround I'm using is to create a dictionary within the state variables called `relevant_vars`

Simple example below:

```
@action(read=["relevant_vars", "task", "prompt"], writes=["plan_prompt"])
def create_plan_prompt(state: State) -> Tuple[dict, State]:
    if state["task"] == "Clear the table":
        plan_prompt = f"""Create a plan for a robot to remove the following objects from the table:
            {state["relevant_vars"]["objects_on_table"]}
            Do not add any extra steps.
        """
    else:
        plan_prompt = "Unknown task"
    result = {"plan_prompt": plan_prompt}
    return result, state.update(**result)
    ```

**Comment by elijahbenizzy** [comment]
This has been implemented, and will be released shortly! See #179
